File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/Tempgbox.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 * TempGBox 渠道 — https://tempgbox.net/api/proxy
 * JSON API，响应以 base64 编码嵌入 HTML data-x 属性中。
 */
public final class Tempgbox {

    private static final String CHANNEL = "tempgbox";
    private static final String API_URL = "https://tempgbox.net/api/proxy";
    private static final String ORIGIN = "https://tempgbox.net";

    private Tempgbox() {
    }

    private static Map<String, String> defaultHeaders(String deviceId) {
        String ip = randomIp();
        Map<String, String> h = new LinkedHashMap<>();
        h.put("Accept", "text/html,application/json");
        h.put("Content-Type", "application/json");
        h.put("Origin", ORIGIN);
        h.put("Referer", ORIGIN + "/");
        h.put("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                + "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36");
        h.put("X-Device-ID", deviceId);
        h.put("X-Forwarded-For", ip);
        h.put("X-Real-IP", ip);
        h.put("X-Originating-IP", ip);
        return h;
    }

    private static String randomDeviceId() {
        byte[] bytes = new byte[32];
        ThreadLocalRandom.current().nextBytes(bytes);
        StringBuilder sb = new StringBuilder(64);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xFF));
        }
        return sb.toString();
    }

    private static String randomIp() {
        ThreadLocalRandom r = ThreadLocalRandom.current();
        return (r.nextInt(1, 255)) + "." + (r.nextInt(1, 255)) + "."
                + (r.nextInt(1, 255)) + "." + (r.nextInt(1, 255));
    }

    private static JsonObject decodePayload(String html) {
        // 查找 data-x="..." 或 data-x='...'
        String marker = "data-x=\"";
        int start = html.indexOf(marker);
        String quote = "\"";
        if (start < 0) {
            marker = "data-x='";
            start = html.indexOf(marker);
            quote = "'";
        }
        if (start < 0) {
            throw new RuntimeException("tempgbox: missing encoded response payload");
        }
        start += marker.length();
        int end = html.indexOf(quote, start);
        if (end < 0) {
            throw new RuntimeException("tempgbox: malformed encoded response payload");
        }
        String encoded = html.substring(start, end);
        String decoded = new String(Base64.getDecoder().decode(encoded), StandardCharsets.UTF_8);
        JsonElement parsed = Json.parse(decoded);
        if (parsed == null || !parsed.isJsonObject()) {
            throw new RuntimeException("tempgbox: payload 解析失败");
        }
        return parsed.getAsJsonObject();
    }

    private static JsonObject postProxy(String route, String deviceId, Map<String, Object> payload) {
        String url = API_URL + "?route=" + ProviderUtil.urlEncode(route);
        HttpResult resp = HttpClient.post(url, Json.serialize(payload), "application/json",
                defaultHeaders(deviceId));
        JsonObject data = decodePayload(resp.getBody());
        if (!resp.isOk()) {
            String reason = ProviderUtil.firstNonEmpty(
                    Json.str(data, "detail"), Json.str(data, "error"), Json.str(data, "message"));
            throw new RuntimeException("tempgbox " + route + " failed: " + resp.getStatusCode() + " " + reason);
        }
        return data;
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        String deviceId = randomDeviceId();
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("variant", "googlemail");
        JsonObject data = postProxy("generate", deviceId, payload);

        JsonElement aliasEl = data.get("alias");
        if (aliasEl == null || !aliasEl.isJsonObject()) {
            throw new RuntimeException("tempgbox: missing alias in response");
        }
        JsonObject alias = aliasEl.getAsJsonObject();
        String email = ProviderUtil.firstNonEmpty(Json.str(alias, "email"), Json.str(alias, "alias"));
        if (email.isEmpty()) {
            throw new RuntimeException("tempgbox: missing email");
        }

        String createdAt = Json.str(alias, "created_at");
        String expiresAt = Json.str(alias, "expires_at");
        return new EmailInfo(CHANNEL, email, deviceId,
                expiresAt.isEmpty() ? null : null, createdAt.isEmpty() ? null : createdAt);
    }

    /**
     * 获取邮件列表。
     *
     * @param token    device ID
     * @param email    邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        if (token == null || token.isBlank()) {
            throw new IllegalArgumentException("tempgbox: missing device id");
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("email", email != null ? email : "");
        JsonObject data = postProxy("inbox", token, payload);

        JsonElement msgsEl = data.get("messages");
        if (msgsEl == null || !msgsEl.isJsonArray()) return new ArrayList<>();

        List<Email> result = new ArrayList<>();
        for (JsonElement msgEl : msgsEl.getAsJsonArray()) {
            if (!msgEl.isJsonObject()) continue;
            JsonObject raw = msgEl.getAsJsonObject();
            Map<String, Object> flat = Json.toDict(msgEl);
            // 映射字段
            flat.put("from", ProviderUtil.firstNonEmpty(Json.str(raw, "from"), Json.str(raw, "sender")));
            flat.put("text", ProviderUtil.firstNonEmpty(Json.str(raw, "text"), Json.str(raw, "body_text")));
            flat.put("html", ProviderUtil.firstNonEmpty(Json.str(raw, "html"), Json.str(raw, "body_html")));
            flat.put("date", ProviderUtil.firstNonEmpty(Json.str(raw, "date"), Json.str(raw, "received_at")));
            flat.put("messageId", ProviderUtil.firstNonEmpty(
                    Json.str(raw, "messageId"), Json.str(raw, "message_id"), Json.str(raw, "id")));
            result.add(Normalizer.normalizeEmail(flat, email));
        }
        return result;
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/Tempgbox.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Http
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*
import java.nio.charset.StandardCharsets
import java.util.Base64
import kotlin.random.Random

/**
 * TempGBox 渠道（tempgbox.net/api/proxy）。
 *
 * JSON API，响应以 base64 编码嵌入 HTML data-x 属性中。token 为 device id。
 */
object Tempgbox : Provider {

    private const val CHANNEL = "tempgbox"
    private const val API_URL = "https://tempgbox.net/api/proxy"
    private const val ORIGIN = "https://tempgbox.net"

    private fun defaultHeaders(deviceId: String): Map<String, String> {
        val ip = randomIp()
        return mapOf(
            "Accept" to "text/html,application/json",
            "Content-Type" to "application/json",
            "Origin" to ORIGIN,
            "Referer" to "$ORIGIN/",
            "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
                "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36",
            "X-Device-ID" to deviceId,
            "X-Forwarded-For" to ip,
            "X-Real-IP" to ip,
            "X-Originating-IP" to ip,
        )
    }

    private fun randomDeviceId(): String {
        val bytes = ByteArray(32)
        Random.nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it.toInt() and 0xFF) }
    }

    private fun randomIp(): String =
        "${Random.nextInt(1, 255)}.${Random.nextInt(1, 255)}.${Random.nextInt(1, 255)}.${Random.nextInt(1, 255)}"

    /** 从 HTML 的 data-x 属性中解码 base64 payload。 */
    private fun decodePayload(html: String): JsonObject {
        var marker = "data-x=\""
        var start = html.indexOf(marker)
        var quote = "\""
        if (start < 0) {
            marker = "data-x='"
            start = html.indexOf(marker)
            quote = "'"
        }
        if (start < 0) throw RuntimeException("tempgbox: 缺少编码响应 payload")
        start += marker.length
        val end = html.indexOf(quote, start)
        if (end < 0) throw RuntimeException("tempgbox: 编码响应 payload 格式错误")
        val encoded = html.substring(start, end)
        val decoded = String(Base64.getDecoder().decode(encoded), StandardCharsets.UTF_8)
        return Http.json.parseToJsonElement(decoded) as? JsonObject
            ?: throw RuntimeException("tempgbox: payload 解析失败")
    }

    /** 通过 proxy 发起 POST 请求。 */
    private suspend fun postProxy(route: String, deviceId: String, payload: JsonObject): JsonObject {
        val url = "$API_URL?route=${ProviderUtil.urlEncode(route)}"
        val resp = ProviderUtil.httpPost(url, payload.toString(), "application/json", defaultHeaders(deviceId))
        val data = decodePayload(resp.body)
        if (!resp.isOk) {
            val reason = ProviderUtil.firstNonEmpty(
                jsonStr(data, "detail"), jsonStr(data, "error"), jsonStr(data, "message"),
            )
            throw RuntimeException("tempgbox $route failed: ${resp.statusCode} $reason")
        }
        return data
    }

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val deviceId = randomDeviceId()
        val payload = buildJsonObject { put("variant", "googlemail") }
        val data = postProxy("generate", deviceId, payload)
        val alias = data["alias"] as? JsonObject
            ?: throw RuntimeException("tempgbox: 响应中缺少 alias")
        val email = ProviderUtil.firstNonEmpty(jsonStr(alias, "email"), jsonStr(alias, "alias"))
        if (email.isEmpty()) throw RuntimeException("tempgbox: 缺少邮箱地址")
        val createdAt = jsonStr(alias, "created_at")
        return EmailInfo(email = email, channel = CHANNEL, token = deviceId, createdAt = createdAt)
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val token = info.token
        if (token.isBlank()) throw RuntimeException("tempgbox: 缺少 device id")
        val email = info.email
        val payload = buildJsonObject { put("email", email) }
        val data = postProxy("inbox", token, payload)
        val msgs = data["messages"] as? JsonArray ?: return emptyList()
        return msgs.mapNotNull { it as? JsonObject }.map { raw ->
            val flat = mapOf(
                "id" to ProviderUtil.firstNonEmpty(
                    jsonStr(raw, "messageId"), jsonStr(raw, "message_id"), jsonStr(raw, "id"),
                ),
                "from" to ProviderUtil.firstNonEmpty(jsonStr(raw, "from"), jsonStr(raw, "sender")),
                "to" to email,
                "subject" to jsonStr(raw, "subject"),
                "body" to ProviderUtil.firstNonEmpty(
                    jsonStr(raw, "text"), jsonStr(raw, "body_text"),
                    jsonStr(raw, "html"), jsonStr(raw, "body_html"),
                ),
                "date" to ProviderUtil.firstNonEmpty(jsonStr(raw, "date"), jsonStr(raw, "received_at")),
            )
            Normalize.fromMap(flat, email)
        }
    }
}

/** 从 JSON 对象读取字符串字段，缺失返回空串（文件私有）。 */
private fun jsonStr(obj: JsonObject, key: String): String =
    (obj[key] as? JsonPrimitive)?.contentOrNull ?: ""
